exercises/nucleotide-count/canonical-data.json

```diff
@@ -39,12 +71,38 @@
           }
         },
         {
-          "description": "strand with invalid nucleotides",
+          "description": "returns the count of nucleotides in a strand containing mixed-case nucleotides.",
           "property": "nucleotideCounts",
-          "strand": "AGXXACT",
+          "strand": "aGcTtTtCaTtCtGaCtGcAaCgGgCaAtAtGtCtCtGtGtGgAtTaAaAaAaGaGtGtCtGaTaGcAgC",
           "expected": {
-            "error": "Invalid nucleotide in strand"
+            "A": 20,
+            "C": 12,
+            "G": 17,
+            "T": 21
           }
+        },
+        {
+          "comments" : [
+            "This should call the property/function at least two times and check if it returns the same",
+            "result. This prevents errors where the call aggregates (sums up) the counts of each call."
+          ],
+          "description": "returns the same amounts for consecutive nucleotideCounts calls.",
+          "property": "nucleotideCounts",
+          "strand": "AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC",
+          "expected" : true
+        },
+        {
+          "description": "handles incorrect inputs.",
+          "cases": [
+            {
+              "description": "throws error if DNA input does not include anything but ACGT-characters.",
+              "property": "nucleotideCounts",
+              "strand": "AGUUÄ1+ :'A'}CT",
```

Comment:
can you confirm whether or not Ä is in ASCII, given (I think) a declaration of intent in https://github.com/exercism/problem-specifications/pull/895#discussion_r139198420

> I only included error handling of ASCII chars.

(Let me know if I simply misinterpreted the declaration)

Reply:
Yep, `Ä` is ASCII: :-)
http://www.theasciicode.com.ar/extended-ascii-code/letter-a-umlaut-diaeresis-a-umlaut-lowercase-ascii-code-132.html

(Btw: `ß` too, wich was part of the Unicode-discussion ;-) )

Reply:
`Ä` is an upper ASCII character.  Is the intent to restrict to printable characters?  `Ä` (142) is not considered a printable character.

Reply:
@rpottsoh Printable in what context?
In ASCII everything is considered printable from 32-255 (excluding 127).

In the context of this exercise only aA,cC,gG,tT is considered valid.
In another exercise (e.g. pangram) it might be everything from a-zA-Z.
In other cases even everything technically printable, might be valid. It really depends.

The issue discussed here was just, whether it is a Unicode-exclusive character like `℣` or `💩`, because there is consensus not to check them.

Reply:
Chars 32 through 126 are considered printable.  128 through 255 are considered extended.

Reply:
> Yep, `Ä` is ASCII: :-)
>
> http://www.theasciicode.com.ar/extended-ascii-code/letter-a-umlaut-diaeresis-a-umlaut-lowercase-ascii-code-132.html

In this file I see the two bytes 0xc3 0x84, which is in accordance with what you say since 0x84 is 132. Do you know what the 0xc3 is for? I couldn't figure it out, hoping you can help

Reply:
I don't see how this change is any better than the original test case.  In this exercise only ACGT are valid.  How is including extended ASCII characters to `"strand"` improving this test case?

Reply:
@petertseng im a bit confused. Where do you see characters 0x84 and 0xc3?
0x84 = 132 is the lowercase ä. In the file you should see the uppercase Ä = 142 instead.
And 0xc3 = 195 is `├`. Which I don't see here.

@rpottsoh The intention was to provide some random invalid characters from the ASCII space instead of just an X. But for this I'd like to refer you to the separate discussion thread at #902.

Reply:
> I'm a bit confused. Where do you see the characters 0x84 and 0xc3?

```
$ xxd canonical-data.json | grep AGUU
00000cf0: 6e64 223a 2022 4147 5555 c384 312b 203a  nd": "AGUU..1+ :
```

Let me know if you get a different result from doing the same on this file

Reply:
That's utf8 for unicode u00c4 which is capital umlaut a

Reply:
Jep. You are assuming ASCII encoding of the file, but the standard is UTF8. So it's reasonable why you see these char numbers. They are still ASCII compliant tough, because the usual functions like ```charAt()``` and others usually convert to ASCII numbers by default.

Reply:
OK, and this is as you intended, correct?

Reply:
OK, and this (the fact that the file containing 0xc3 0x84) is as you intended, correct? As you can see, all I needed was to be sure it matches your intent.

Reply:
It was strange to me because I thought https://github.com/exercism/problem-specifications/pull/895#discussion_r139296590

> In the file you should see the uppercase Ä = 142 instead.

meant that one would expect to see byte 142 (0x8e) in the file, but that byte does not appear.

Also note that because the bytes are 0x84 and 0xc3, they would appears in the search I used in https://github.com/exercism/problem-specifications/issues/428#issuecomment-258366030

> `git grep -Pnl "[\x80-\xFF]"`

This may mean that I misunderstood the correct definitions somewhere. As you can see, I'm far from being an expert on this matter, so it's quite possible someone needs to set me straight.

Reply:
I am no expert as well, but I think you might have a tripped upon a fallacy in encoding: You can't grep UTF8 encoded files assuming ASCII char numbers. You have to grep them with Unicode UTF8 encoding in mind.

Let me explain:
The json file here is encoded in Unicode UTF8. Which means that the symbol `Ä` will be encoded to bytecode as `0xC3 0x84`.

What you have assumed is that 0xC3 and 0x84 are ASCII-encoded characters. This means you interpreted the bytecode as ASCII, which translates to 0xc3 = 195 = `├` and 0x84 = 132 = `ä`. 
Obviously this is not what the file shows, just look at the file in your browser or any text editor.

Another way you could have interpret the bytecode wrongly is by assuming ANSI encoding. This would translate to `Ã` and `„`. Or Mac encoding: `√` and `Ñ`. Or EBDCID encoding: `C` and `d`.

Now let me show you how the bytecode looks, if I encode the file itself in ASCII. 
This is what you just xxd'ed with the UTF8 encoded file:
`6E 64 22 3A 20 22 41 47 55 55 C3 84 31 2B 20 3A`
This is what you xxd with the file encoded in ASCII:
`6E 64 22 3A 20 22 41 47 55 55 8E 31 2B 20 3A 27` 
This is what you xxd with the file encoded in ANSI:
`6E 64 22 3A 20 22 41 47 55 55 C4 31 2B 20 3A 27` 

Notice that most codes are the same. This is due to UTF8s backwards compatibility with ASCII. However, `Ä` is greater than the ASCII representation of 127, so it needs two bits in UTF8:
```
6E 64 22 3A 20 22 41 47 55 55 C3 84 31 2B 20 3A      // UTF8
6E 64 22 3A 20 22 41 47 55 55 8E    31 2B 20 3A 27   // ASCII
6E 64 22 3A 20 22 41 47 55 55 C4    31 2B 20 3A 27   // ANSI
``` 


The encoding is very important. It is the basics upon how the bytecode will be interpreted. In UTF8 case here, C3 and 84 are actually interpreted as one group of bytes to `Ä`. But this does not mean you can just grep for two-byte values now. UTF8 is actually even more complicated, because depending on what symbol is encoded, the number of bytes varies. See here for a nice explanation:
http://www.fileformat.info/info/unicode/utf8.htm
To use xxd, you actually have to couple it with an UTF8-aware text viewer, like this post explains:
https://askubuntu.com/a/848696

Reply:
> The encoding is very important.

Yupp, and since we don't even know if a track will **encode** the data as a string or something else (especially functional languages may tend to wrap it in an ADT), we shouldn't even think about creating invalid input that might not be even representable.

Also there might be languages which require your source code in certain encodings, which can't even represent `Ä`.

Or transcoding the UTF-8 JSON string into the encoding required by the targetted compiler might be problematic.

So we have a plenty of reasons to stick to 7bit ASCII, an encoding which is even a direct subset of **all** today important encodings.

Reply:
> You can't grep UTF8 encoded files assuming ASCII char numbers

I'm sorry, I don't understand. The specific grep I performed (`grep AGUU`) doesn't affect the bytes in the files; it simply makes it so that the only line that xxd displays is the line with the bytes we are discussing, since I grepped for the characters immediately before. If I need to use grep a different way, can I ask if you know any proper flags I need to pass to grep?

Alternatively, if you were referring to the other grep I mentioned (`git grep -Pnl "[\x80-\xFF]"`) then I acknowledge that the specific characters found by it will be dependent on encoding; if we need a different pattern, I will discuss and find in #428.

> This is what you xxd with the file encoded in ANSI
> To use xxd, you actually have to couple it with an UTF8-aware text viewer

I'm sorry, I don't understand either of these two assertions. I thought xxd simply shows me the bytes in the file. Can I ask you do one of the two following?

1. Show me different flags xxd on canonical-data.json which demonstrates the three possibilities you report (reproduced below)
2. OR show me any other way of reading the exact bytes in canonical-data.json that demonstrates that the byte(s) that you desire exist in the file?

```
6E 64 22 3A 20 22 41 47 55 55 C3 84 31 2B 20 3A      // UTF8
6E 64 22 3A 20 22 41 47 55 55 8E    31 2B 20 3A 27   // ASCII
6E 64 22 3A 20 22 41 47 55 55 C4    31 2B 20 3A 27   // ANSI
```

What I thought I did was to report to you the bytes as they lie in this file on my hard drive as they were delivered to me by Git. Is that not what I did?

Reply:
> Show me
> Show me

Why am I asking to be shown? Simply so that I know that you are OK with the bytes that exist in this file, and they match your intention.

Reply:
Sorry for the grep confusion. I thought you xxd'ed and grep'ed afterwards. I am on Windows, so I am not that firm with unix commands.

> Alternatively, if you were referring to the other grep I mentioned (git grep -Pnl "[\x80-\xFF]") then I acknowledge that the specific characters found by it will be dependent on encoding; if we need a different pattern, I will discuss and find in #428.
Jup, that's what I mean.

Let me say it in other words:
When you view the bytecode of the file now as-is from git, it will show you what you already reported:
`6E 64 22 3A 20 22 41 47 55 55 -> C3 84 <- 31 2B 20 3A`
However, the file has no separate ASCII characters C3 and 84 in it, but has a single character C384, because it is UTF8-encoded.

If I would have encoded the same text content of the file in ASCII, though, and sent it to you, you *would* see the following byte code:
`6E 64 22 3A 20 22 41 47 55 55 -> 8E <- 31 2B 20 3A 27`
This was just included by me for demonstration purposes.

Reply:
> However, the file has no separate ASCII characters C3 and 84 in it, but has a single character C384, because it is UTF8-encoded.

I agree with this statement. The file does not contain the characters 0xC3 and 0x84. It contains the **bytes** 0xC3 and 0x84, which together represent Unicode codepoint 196 (Feel free to check my math: 0xc3 is 110**00011** and 0x84 is 10**000100** so we concatenate 00011 and 000100 to get 11000100 which is 196). So we seem to be on the same page with this, good.

> If I would have encoded the same text content of the file in ASCII, though, and sent it to you, you *would* see the following byte code:
> `6E 64 22 3A 20 22 41 47 55 55 -> 8E <- 31 2B 20 3A 27`

Got it.

Reply:
Yep, we are on the same page. :-)

Reply:
I didn't see that you had edited your post.

> > All test cases should only use ASCII characters
> 
> In my interpretation this meant all ASCII, including the extended ASCII (8-bit).
> In contrast, I think you interpreted it as the 7-bit ASCII only, didn't you?

It is true that given that I grepped for bytes 0x80 to 0xFF that I looked for files that had characters outside of 7-bit ASCII. I wouldn't know for sure what the original text meant, but I know I got no objections at the time, so that is moderate evidence in favour of 7-bit ASCII. If it does in fact mean 7-bit ASCII, it would imply needing to remove the Ä. I'll leave that decision to the interested parties.